app_integrations/apps/duo.py

```diff
@@ -0,0 +1,209 @@
+
+        auth_string = '\n'.join([formatted_date, 'GET', hostname,
+                                 self._endpoint(), urllib.urlencode(params)])
+
+        try:
+            signature = hmac.new(self._config['auth']['secret_key'],
+                                 auth_string, hashlib.sha1)
+        except TypeError:
+            LOGGER.exception('Could not generate hmac signature')
+            return False
+
+        # Format the basic auth with integration key and the hmac hex digest
+        basic_auth = ':'.join([self._config['auth']['integration_key'],
+                               signature.hexdigest()])
+
+        return {
+            'Date': formatted_date,
+            'Authorization': 'Basic {}'.format(b64encode(basic_auth)),
+            'Host': hostname
+        }
+
+    def _gather_logs(self):
+        """Gather the authentication log events."""
+        hostname = self._config['auth']['api_hostname']
+        full_url = 'https://{hostname}{endpoint}'.format(
+            hostname=hostname,
+            endpoint=self._endpoint()
+        )
+
+        return self._get_auth_logs(hostname, full_url)
+
+    def _get_auth_logs(self, hostname, full_url):
+        """Get all logs from the endpoint for this timeframe
+
+        Returns:
+            [
+                {
+                    'timestamp': <int:unix timestamp>,
+                    'device': <str:device>,
+                    'username': <str:username>,
+                    'factor': <str:factor>,
+                    'result': <str:result>,
+                    'ip': <str:ip address>,
+                    'new_enrollment': <bool:if event corresponds to enrollment>,
+                    'integration': <str:integration>,
+                    'location': {
+                        'state': '<str:state>',
+                        'city': '<str:city>',
+                        'country': '<str:country>'
+                    }
+                }
+            ]
+        """
+        # Get the last timestamp and add one to it to avoid duplicates
+        # Sanity check mintime as unix timestamp, then transform to string
+        params = {'mintime': str(int(self._last_timestamp + 1))}
+
+        # Contstruct the headers for this request. Every request must be signed
+        headers = self._generate_auth(hostname, params)
+        if not headers:
+            return False
+
+        # Perform the request and get the list of logs
+        response = requests.get(full_url, headers=headers, params=params)
+
+        if not self._check_http_response(response):
+            return False
+
+        logs = response.json()['response']
+
+        # Get the timestamp from the latest event. Duo produces these sequentially
+        # so we can just extract the timestamp from the last item in the list
+        self._last_timestamp = logs[-1]['timestamp']
+
+        # Check if the max amount of logs was returned with this request. If the value
```

Comment:
> If the value is not the max, then we are done polling logs for this timeframe`

Did you mean to say `If the value is the max`?

Reply:
Nope - it's meant to be how it's worded. Duo's only indicator of there being _more_ logs to pull is whether or not the **max** amount of logs possible was returned with a request. So, if less than the max was returned, this means we've collected all the logs up to this timeframe.